### PR TITLE
Fix for dropped/extra comments

### LINKED ISSUE #2546: TypeScript removing comments where it shouldn't
comment.ts

```
if (false) {
    //I will be removed
}
//I will be removed
else {
    //I will not be removed
    var tmp;
    //I will be removed
}

try{
    //I will not be removed
    var tmp;
    //I will be removed
}
//I will be removed
catch(e){
    //I will not be removed
    var tmp;
    //I will be removed
}
```

comment.js

```
if (false) {
}
else {
    //I will not be removed
    var tmp;
}
try {
    //I will not be removed
    var tmp;
}
catch (e) {
    //I will not be removed
    var tmp;
}
```

Are there any additional flags that can be set for TS to keep _all_ comments?

[Same issue from CodePlex](https://typescript.codeplex.com/workitem/2573)

### PR DESCRIPTION
**Pull request by Myste**
Fixes #2546, #6069, #6982, #7227.

Fix for #7227 is trivial - leading comments are not emitted for arrow function.

#2546 mentions two problems: comments between if block/else block and comments inside blocks (like #6069, #6982). Dropped comments were not attached to any node, so I added additonal logic in emitIfStatement & emitBlock to emit them.

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -2863,6 +2889,12 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, ge
                 if (node.kind === SyntaxKind.ModuleBlock) {
                     emitTempDeclarations(/*newLine*/ true);
                 }
+
+                if (!!node.statements.length) {
+                    const lastStatementNode = node.statements[node.statements.length - 1];
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: We have a `lastOrUndefined` for these

**Comment by martine** [comment]
What's the status of this PR?  Does it need more work or does it need review?

**Comment by RussKie** [comment]
I've scrolled through the change, but didn't see the answer.

Does it fix [this issue](http://www.typescriptlang.org/play/#src=%2F%2F%20the%20export%20annotations%0Aexport%20function%20bug()%20%7B%0A%09%2F%2F%20hello%0A%09%0A%09%2F%2F%20another%3F%0A%09%0A%09%2F%2F%20return%20annotation%0A%09return%20%7B%0A%09%09%2F%2F%20return%20object%20annotation%0A%09%09outer%3A%20%7B%0A%09%09%09%2F%2F%20more!%0A%09%09%09inner1%3A%20%7B%20%2F%2F%20lost_comment1%0A%09%09%09%09%2F%2F%20and%20more%0A%09%09%09%09%0A%09%09%09%09p1%3A%20'v1'%2C%20%2F%2F%20lost_comment2%0A%09%09%09%09p2%3A%20'v2'%2C%20%2F%2F%20lost_comment3%0A%09%09%09%09p3%3A%20'v3'%20%20%2F%2F%20last%20comment%0A%09%09%09%7D%20%2F%2F%20here%3F%0A%09%09%7D%20%2F%2F%20and%20here%3F%0A%09%7D%3B%20%2F%2F%20here%20too%3F%0A%09%0A%09%2F%2F%20and%20finally%20here%0A%7D) (i.e. `lost_commentX` are gone)?

Thanks

**Comment by RyanCavanaugh** [comment]
@mhegazy thoughts on this one? I like the emit changes

**Comment by mhegazy** [comment]
👍

**Comment by mhegazy** [comment]
we will need to port this to the transforms branch. @yuit  can you take care of that?

**Comment by mhegazy** [comment]
Just noticed that this was not merged in when i made the comment. Extremely sorry for the delay. unfortunately with https://github.com/Microsoft/TypeScript/issues/5595 the emitter has fundamentally changed. the bugs do exist however.  We would appreciate it if you could port this change to the new emitter. 
Since it is not possible to manually merge this change because of the rewrite, I will go ahead and close this PR. sorry again for the delay, and hopefully future PRs would not need to wait this long to get merged in.

**Pull request closed without merging**